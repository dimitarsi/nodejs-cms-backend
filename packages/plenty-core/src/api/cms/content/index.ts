import { FastifyPluginCallback } from "fastify"
import getAll from "./actions/getAll"
import getById from "./actions/getById"
import createContents from "./actions/create"
import updateContent from "./actions/update"
import deleteContent from "./actions/delete"
import projectAccess from "@middleware/projectAccess"

const contentController: FastifyPluginCallback = (instance, options, done) => {
  instance.register((instance, _, done) => {
    projectAccess(instance, { accessLevel: "readOrUp" })

    getAll(instance)
    getById(instance)

    done()
  })

  instance.register((instance, _, done) => {
    projectAccess(instance, { accessLevel: "writeOrUp" })

    createContents(instance)
    updateContent(instance)
    deleteContent(instance)

    done()
  })

  done()
}

export default contentController
