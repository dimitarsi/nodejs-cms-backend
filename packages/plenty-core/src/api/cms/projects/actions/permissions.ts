import { FastifyInstance, RouteShorthandOptions } from "fastify"

const routeOptions: RouteShorthandOptions = {}

export default function updatePermissions(instance: FastifyInstance) {
  instance.post<{
    Params: {
      projectId: string
      userId: string
    }
  }>(
    "/projects/:projectId/permissions/:userId",
    routeOptions,
    async (request, reply) => {}
  )
}
