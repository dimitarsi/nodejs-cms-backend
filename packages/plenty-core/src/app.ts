import "./config"
import cmsAPI from "~/api/cms"
import "~/middleware/auth"
import fastify from "fastify"
import repo from "~/repo"
import publicApiAuth from "~/api/public"
import fastifyPlugin from "fastify-plugin"
import cmsApiSchema from "~/schema/cmsAPISchema"
import services from "~/services/register"
import fastifySensible from "@fastify/sensible"
import successResponses from "./plugins/successResponses"

const app = fastify({
  logger: {
    level: "info",
    file: "./server.log",
  },
})

app.register(fastifyPlugin(successResponses))
app.register(fastifyPlugin(fastifySensible))
app.register(fastifyPlugin(cmsApiSchema))
app.register(fastifyPlugin(repo), {
  dbName: process.env.DB_NAME || "plenty_cms",
})

app.register(fastifyPlugin(services))

app.register(cmsAPI)
app.register(publicApiAuth)

app.get("/health", (_r, reply) => reply.status(200).send("OK"))

export default app
