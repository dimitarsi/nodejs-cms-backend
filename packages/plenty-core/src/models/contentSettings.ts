import type { ObjectId } from "mongodb"

export default interface ContentSettings {
  _id: ObjectId | string
  contentId: ObjectId | string
  previewUrl: {
    slug: string
  }
}
