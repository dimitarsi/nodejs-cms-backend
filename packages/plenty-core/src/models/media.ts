export interface MediaDocument {
  fileId: string
  ext: string
  path: string
  mimetype: string
  absolutePath: string
  location: string
  // filetype: string
  originalName: string
  // size: string
}
