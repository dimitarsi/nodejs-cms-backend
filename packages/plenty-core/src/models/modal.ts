import { ObjectId } from "mongodb"

export interface BaseModel {
  _id?: string | ObjectId
  createdOn: Date
  updatedOn: Date
  active: boolean
}
