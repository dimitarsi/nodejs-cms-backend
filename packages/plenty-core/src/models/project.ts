import type { ObjectId } from "mongodb"

export interface DomainSettings {
  host: string
  defaultLocale: string
  supportedLanguages: string[]
}

export interface Project {
  _id: ObjectId | string
  name: string
  active: boolean
  domains: DomainSettings[]
  owner: ObjectId | string // Ref to user
}
