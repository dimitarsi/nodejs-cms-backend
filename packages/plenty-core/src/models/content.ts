import { ObjectId } from "mongodb"
import { BaseModel } from "./modal"
import { ContentType } from "./contentType"

type FileId = string
type RefId = string

export type TextDataType = {
  type: "text"
  value: string
}

export type FileDataType = {
  type: "files"
  value: FileId
}

export type RefDataType = {
  type: "ref"
  value: {
    refId: RefId
    type: string
  }
}

export type ComponentDataType = {
  type: "component"
  componentId: string | ObjectId
  value: Record<string, DataTypes>
}

export type DataTypes =
  | TextDataType
  | FileDataType
  | RefDataType
  | ComponentDataType

export interface BaseContentData<D = any, C = ContentType> {
  name: string
  slug: string
  config: C
  children: BaseContentData[]
  outdated?: boolean
  data: D
}

export interface TopLevelContentData {
  isFolder: boolean
  folderLocation: string
  folderTarget: string
  configId?: ObjectId | string
  projectId: ObjectId | string
}

export type CreateContentPayload = TopLevelContentData &
  Omit<BaseContentData, "config">

export type ComputedContentFields = {
  folderDepth: number
}

export type UpdateContentPayload = CreateContentPayload & ComputedContentFields

export type Content = BaseModel & UpdateContentPayload

export type ContentWithConfig<T = ContentType> = Content & {
  config: T
}
