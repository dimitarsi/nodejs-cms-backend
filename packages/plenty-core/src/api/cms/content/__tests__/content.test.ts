import { describe, test, beforeAll, afterAll, afterEach, expect } from "vitest"
import { MongoClient, ObjectId } from "mongodb"
import supertest from "supertest"
import app from "~/app"
import { CreateContentPayload } from "~/models/content"
import accessTokens from "~/repo/accessTokens"
import permissions from "~/repo/permissions"

const createContent = (): CreateContentPayload => ({
  name: "foo",
  slug: "foo",
  isFolder: false,
  folderLocation: "/",
  folderTarget: "",
  children: [],
  data: null,
  projectId: new ObjectId(),
})

describe("Content", async () => {
  const mongoClient = new MongoClient(
    process.env.MONGO_URL || "mongodb://root:example@localhost:27017"
  )
  const db = await mongoClient.db(process.env.DB_NAME)
  const accessTokensRepo = accessTokens(db)
  const permissionsRepo = permissions(db)
  const projectId = new ObjectId().toString()
  const adminUserId = new ObjectId()
  const nonAdminUserId = new ObjectId()

  const CONTENT_API_ADMIN_TOKEN = "content-API-admin-token"
  const CONENT_API_NON_ADMIN_TOKEN = "content-API-non-admin-token"

  // Seed tokens
  beforeAll(async () => {
    await Promise.all([
      accessTokensRepo.findOrCreateAccessToken(
        adminUserId,
        CONTENT_API_ADMIN_TOKEN
      ),
      accessTokensRepo.findOrCreateAccessToken(
        nonAdminUserId,
        CONENT_API_NON_ADMIN_TOKEN
      ),
      permissionsRepo.setAdminUser(adminUserId, projectId, "grant"),
      permissionsRepo.setAdminUser(nonAdminUserId, projectId, "revoke"),
    ])
  })

  afterAll(async () => {
    await accessTokensRepo.deactivateToken(CONTENT_API_ADMIN_TOKEN)
    await accessTokensRepo.deactivateToken(CONENT_API_NON_ADMIN_TOKEN)

    await app.close()
    await mongoClient.close()
  })

  afterEach(async () => {
    await db.collection("contents").deleteMany({})
  })

  describe("Needs Authentication", () => {
    test("GET /:projectId/contents", async () => {
      await app.ready()

      await supertest(app.server).get(`/${projectId}/contents`).expect(403)
    })

    test("GET /:projectId/contents/:id", async () => {
      await app.ready()

      await supertest(app.server)
        .get(`/${projectId}/contents/foobar`)
        .expect(403)
    })

    test("POST /projectId/contents", async () => {
      await app.ready()

      await supertest(app.server)
        .post(`/${projectId}/contents`)
        .send(createContent())
        .expect(403)
    })

    test("PATCH /:projectId/contents/foobar", async () => {
      await app.ready()

      await supertest(app.server)
        .patch(`/${projectId}/contents/foobar`)
        .send({
          hello: "world",
        })
        .expect(403)
    })

    test("DELETE /:projetId/contents/foobar", async () => {
      await app.ready()

      await supertest(app.server)
        .delete(`/${projectId}/contents/foobar`)
        .expect(403)
    })
  })

  describe("Only Admin users can create, update, delete and inspect content types", () => {
    describe("As Admin user", () => {
      test("GET /:projectId/contents", async () => {
        await app.ready()

        supertest(app.server)
          .get(`/${projectId}/contents`)
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(200)
      })

      test("POST /:projectId/contents", async () => {
        await app.ready()

        const data = createContent()

        expect(data.isFolder).toBeDefined()

        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()
        expect(resp.body).toHaveProperty("_id")
        expect(resp.body).toHaveProperty("name")
        expect(resp.body).toHaveProperty("slug")
        expect(resp.body).toHaveProperty("isFolder")
        expect(resp.body).toHaveProperty("folderTarget")
        expect(resp.body).toHaveProperty("folderLocation")
        expect(resp.body).toHaveProperty("folderDepth")
        expect(resp.body).toHaveProperty("children")
        expect(resp.body.folderDepth).toBe(0)
        expect(resp.body).toHaveProperty("data")

        const { _id, createdOn, updatedOn, ...body } = resp.body
        const { projectId: bodyProjectId, ...bodyNoProjectId } = body

        expect(bodyProjectId).toBeTruthy()
        expect(bodyNoProjectId).toMatchSnapshot()

        expect(resp.headers["location"]).toMatch(
          new RegExp(`^\/${projectId}\/contents\/(.+)$`)
        )
      })

      test("GET /:projectId/contents/:idOrSlug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()

        const resultFromSlug = await supertest(app.server)
          .get(`/${projectId}/contents/${resp.body.slug}`)
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(200)

        const resultFromId = await supertest(app.server)
          .get(`/${projectId}/contents/${resp.body._id}`)
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(200)

        expect(resultFromId.body._id).toBeDefined()
        expect(resultFromSlug.body._id).toBeDefined()
        expect(resultFromSlug.body).toHaveProperty("children")
        expect(resultFromId.body._id).toBe(resultFromSlug.body._id)
      })

      test("PATCH /:projectId/contents/:idOrSlug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .patch(`/${projectId}/contents/${resp.body.slug}`)
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .send({
            name: "Hello World",
          })
          .expect(200)

        await supertest(app.server)
          .patch(`/${projectId}/contents/${resp.body._id}`)
          .send({
            name: "Hello World 2",
          })
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(200)
      })

      test("PATCH /:projectId/contents/:idOrSlug - changing folder also updates the depth", async () => {
        await app.ready()

        const data = createContent()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(data)
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()
        expect(resp.body.folderDepth).toBe(0)

        const response = await supertest(app.server)
          .patch(`/${projectId}/contents/${resp.body.slug}`)
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .send({
            name: "Hello World",
            folderLocation: "/posts",
          })
          .expect(200)

        expect(response.body.folderDepth).toBe(1)
        expect(response.body.name).toBe("Hello World")
        expect(response.body.slug).toBe(data.slug)
      })

      test("DELETE /:projectId/contents/:id", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .delete(`/${projectId}/contents/${resp.body._id}`)
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(200)
      })

      test("DELETE /:projectId/contents/:slug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .delete(`/${projectId}/contents/${resp.body.slug}`)
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(200)
      })
    })

    describe("As non-Admin user", () => {
      test("GET /:projectId/contents", async () => {
        await app.ready()

        supertest(app.server)
          .get(`/${projectId}/contents`)
          .set("X-Access-Token", CONENT_API_NON_ADMIN_TOKEN)
          .expect(403)
      })

      test("POST /:projectId/contents", async () => {
        await app.ready()

        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONENT_API_NON_ADMIN_TOKEN)
          .expect(403)
      })

      test("GET /:projectId/contents/:idOrSlug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .get(`/${projectId}/contents/${resp.body.slug}`)
          .set("X-Access-Token", CONENT_API_NON_ADMIN_TOKEN)
          .expect(403)

        await supertest(app.server)
          .get(`/${projectId}/contents/${resp.body._id}`)
          .set("X-Access-Token", CONENT_API_NON_ADMIN_TOKEN)
          .expect(403)
      })

      test("PATCH /:projectId/contents/:idOrSlug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .patch(`/${projectId}/contents/${resp.body.slug}`)
          .set("X-Access-Token", CONENT_API_NON_ADMIN_TOKEN)
          .send({
            name: "Hello World",
          })
          .expect(403)

        await supertest(app.server)
          .patch(`/${projectId}/contents/${resp.body._id}`)
          .send({
            name: "Hello World 2",
          })
          .set("X-Access-Token", CONENT_API_NON_ADMIN_TOKEN)
          .expect(403)
      })

      test("DELETE /:projectId/contents/:id", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .delete(`/${projectId}/contents/${resp.body._id}`)
          .set("X-Access-Token", CONENT_API_NON_ADMIN_TOKEN)
          .expect(403)
      })

      test("DELETE /:projectId/contents/:slug", async () => {
        await app.ready()

        // Ensure entity exists
        const resp = await supertest(app.server)
          .post(`/${projectId}/contents`)
          .send(createContent())
          .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
          .expect(201)

        expect(resp.body).toBeDefined()

        await supertest(app.server)
          .delete(`/${projectId}/contents/${resp.body.slug}`)
          .set("X-Access-Token", CONENT_API_NON_ADMIN_TOKEN)
          .expect(403)
      })
    })
  })

  describe("Quering content types", () => {
    test("find content by folder - `/:projectId/contents?folder=/posts`", async () => {
      await app.ready()

      const data = createContent()
      const dataInFolder = createContent()

      await supertest(app.server)
        .post(`/${projectId}/contents`)
        .send(data)
        .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
        .expect(201)

      dataInFolder.slug = "test-folder-query"
      dataInFolder.folderLocation = "/posts"
      // dataInFolder.folderDepth = 2

      await supertest(app.server)
        .post(`/${projectId}/contents`)
        .send(dataInFolder)
        .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
        .expect(201)

      await supertest(app.server)
        .get(`/${projectId}/contents/${dataInFolder.slug}`)
        .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
        .expect(200)

      const response = await supertest(app.server)
        .get(`/${projectId}/contents?folder=/posts`)
        .set("X-Access-Token", CONTENT_API_ADMIN_TOKEN)
        .expect(200)

      expect(response.body.items).toHaveLength(1)
      expect(response.body.pagination.count).toBe(1)
      expect(response.body.items[0].slug).toBe(dataInFolder.slug)
      expect(response.body.items[0].folderDepth).toBe(1)
    })
  })
})
