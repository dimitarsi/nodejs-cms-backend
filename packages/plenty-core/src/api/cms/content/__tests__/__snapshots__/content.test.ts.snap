// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Content > Only Admin users can create, update, delete and inspect content types > As Admin user > POST /:projectId/contents 1`] = `
{
  "active": true,
  "children": [],
  "config": [],
  "data": null,
  "folderDepth": 0,
  "folderLocation": "/",
  "folderTarget": "",
  "isFolder": false,
  "name": "foo",
  "slug": "foo",
}
`;
